### PR TITLE
[frontend] Add the lasso selection tool in graphs (#3109)

### LINKED ISSUE #3109: Implement lasso selection in graph in addition to rectangle selection
## Use case

Implement lasso selection in graph in addition to rectangle selection

### PR DESCRIPTION
**Pull request by Kedae**
Be careful while testing it to use as many items in graph as possible as I had to change the render behavior of the Graph. We could have some corner cases I missed.

### CODE REVIEW

**Comment by Archidoit** [comment]
- Select things with lasso
- Select things with rectangle selection
- Some weird behaviors occur

image

image


---
@Kedae : Done

**Comment by SarahBocognano** [comment]
Except the weird behavior that @Archidoit pointed out, everything seems to work.

File: opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledgeCorrelation.jsx [hunk 1]
```diff
@@ -284,7 +271,9 @@ class ReportKnowledgeCorrelationComponent extends Component {
     );
     this.state = {
       mode3D: R.propOr(false, 'mode3D', params),
-      selectModeFree: R.propOr(false, 'selectModeFree', params),
+      selectRectangleModeFree: R.propOr(false, 'selectRectangleModeFree', params),
+      selectModeFree: !!params.selectModeFree,
```
Comment (richard-julien) [inline opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledgeCorrelation.jsx]: Not useful

File: opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledge.jsx [hunk 2]
```diff
@@ -123,7 +123,7 @@ class ReportKnowledgeComponent extends Component {
     const params = buildViewParamsFromUrlAndStorage(
       props.history,
       props.location,
-      `view-report-knowledge-${props.report.id}`,
+      `view-report-${props.report.id}-knowledge`,
```
Comment (richard-julien) [inline opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledge.jsx]: Impact needs to be checked

File: opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledgeCorrelation.jsx [hunk 3]
```diff
@@ -397,8 +411,24 @@ class ReportKnowledgeCorrelationComponent extends Component {
...
+
   handleToggleSelectModeFree() {
-    this.setState({ selectModeFree: !this.state.selectModeFree }, () => this.saveParameters());
+    document.removeEventListener('mousedown', this.startFreeHandFunction);
+    document.removeEventListener('mouseup', this.stopFreeHandFunction);
+    this.setState({
+      selectModeFree: !this.state.selectModeFree,
+      selectRectangleModeFree: false,
+    }, () => this.saveParameters());
+    if (!this.state.selectModeFree) {
```
Comment (richard-julien) [inline opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledgeCorrelation.jsx]: Can be simplify

File: opencti-platform/opencti-front/src/utils/Graph.js [hunk 4]
```diff
@@ -1198,3 +1198,78 @@ export const parseDomain = (data) => [
...
+      // Invert odd
+      odd = !odd;
+    }
+    j = i;
+  }
+  // If the number of crossings was odd, the point is in the polygon
+  return odd;
+};
+
+// TEST FOR UTILITY
```
Comment (richard-julien) [inline opencti-platform/opencti-front/src/utils/Graph.js]: To be cleaned

**Review by richard-julien — CHANGES_REQUESTED** [review]

**Review by richard-julien — COMMENTED** [review]
Comment

**Comment by richard-julien** [comment]
Looks like sometimes click handler are not correctly canceled.
Like @Archidoit already mentionned.
I also be able to do something like this.

image

**Comment by richard-julien** [comment]
Another question also, do we really want to complete the drag in case of open selection?

--- 
@Kedae I honesly thinks that it is much easier to use and to understand that way. May be we should go this way for a V1, and then update the algorithm if some users struggle with it ? @Jipegien Any thoughts on that ?

File: opencti-platform/opencti-front/src/utils/Graph.js (hunk 4 above)
Comment (Kedae) [reply opencti-platform/opencti-front/src/utils/Graph.js]: Sorry, I told you garbage yesterday. Those methods are actually the ones used to make the lasso work. So I'll simply modify the comments to explain that.

File: opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledgeCorrelation.jsx (hunk 1 above)
Comment (Kedae) [reply opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledgeCorrelation.jsx]: Done

File: opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledge.jsx (hunk 2 above)
Comment (Kedae) [reply opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledge.jsx]: Done, misread the code, reverted those

File: opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledgeCorrelation.jsx (hunk 3 above)
Comment (Kedae) [reply opencti-platform/opencti-front/src/private/components/analysis/reports/ReportKnowledgeCorrelation.jsx]: Done

**Comment by Jipegien** [comment]
@richard-julien @Kedae  about drag completion. As a user,  i will be upset to see my entire and complex selection being cancelled because I miss to cross my drag from 3 pixels. Plus, I think if we autocomplete the drag, it will not bother users that used to complete their drags, as it is straightforward to drag again even if elements are selected.

**Review by SarahBocognano — APPROVED** [review]
Tested after changes ✅

**Review by RomuDeuxfois — APPROVED** [review]

**Pull request merged**